Optimizations for array generation and shrinking

jack-pappas opened the pull request:
Modified Gen functions for generating arbitrary array instances to avoid intermediate conversions to F# lists (and resulting allocations).
Modified shrinker for 1D arrays to avoid intermediate list allocations.
Modified shrinker for strings to use the array shrinker, and subsequently avoid allocating intermediate char lists.

I noticed these allocations when profiling a slow-running FsCheck test I wrote at one point (for a data type which used both arrays and strings). Running ``build.cmd`` before and after this change seems to show the tests running slightly faster, although I haven't repeated it enough times for statistical significance. In any case, tests which heavily use arrays and strings should notice a performance improvement due to eliminating the conversions to/from F# lists (and also due to spending less time in GCs related to those list allocations).

kurtschelfthout commented on src/FsCheck/Gen.fs:
```diff
@@ -330,14 +343,38 @@ module Gen =
     [<CompiledName("Sequence"); CompilerMessage("This method is not intended for use from F#.", 10001, IsHidden=true, IsError=false)>]
-    let sequenceToSeq generators = 
-        generators |> sequence |> map List.toSeq
+    let sequenceToSeq<'T, 'TSource when 'TSource :> IEnumerable<Gen<'T>>> (generators: 'TSource) =
```
This function used to only take a single type argument. So this unfortunately breaks compatibility.

jack-pappas replied:
Ah, didn't think about that - I'll revert this part of the changes. Now that I think about it again, I'm not sure adding the extra type parameter was going to accomplish what I had in mind anyway.

kurtschelfthout commented on src/FsCheck/Arbitrary.fs:
```diff
@@ -555,6 +563,32 @@ module Arb =
+            fromGenShrink (generator,shrinker)
+
+        ///Generate an object - a boxed char, string or boolean value.
+        static member Object() =
```
This Object() method is just moved, right?

jack-pappas replied:
Yes. I'll revert that part of the change. I must've been on autopilot and thinking of reordering things as if they were let-bindings. They're all static members though so there's no need to reorder them.

kurtschelfthout commented on src/FsCheck/Arbitrary.fs:
```diff
@@ -555,6 +563,32 @@ module Arb =
+            let shrinker (s:string) = 
+                    match s with
+                    | null -> Seq.empty
+                    | _ -> s.ToCharArray() |> shrink |> Seq.map (fun chars -> new String(chars))
```
Any other change besides removing the toList? (as a side note, diffs are much easier to understand for me if you don't move things around...)

jack-pappas replied:
No, no change besides that. Removing the toList just makes it use the shrinker for array instead.

kurtschelfthout commented:
Overall, I'm liking this PR!

jack-pappas commented:
I've moved the Arbitrary members back to their original locations and fixed that one incompatibility in Gen. The diff should be smaller and more readable now.
